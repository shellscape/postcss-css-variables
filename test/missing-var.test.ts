import type { Result } from 'postcss';

import { fixture, process } from './helpers';

const tests = {
  'use fallback value if provided with missing variables': 'missing-variable-should-fallback',
  'use fallback variable if provided with missing variables':
    'missing-variable-should-fallback-var',
  'use fallback variable if provided with missing variables calc':
    'missing-variable-should-fallback-calc',
  'use fallback variable if provided with missing variables nested':
    'missing-variable-should-fallback-nested',
  'work with missing variables': 'missing-variable-usage'
};

describe('missing-var', async () => {
  test.each(Object.entries(tests))('%s', async (_desc, file) => {
    const css = await fixture(`missing-var/${file}`);
    const result = await process(css);
    expect(result).toMatchSnapshot();
  });

  test('use string values for `undefined` values, see #22', async () => {
    const css = await fixture('missing-var/missing-variable-usage');
    const result = (await process(css, void 0, true)) as Result;
    const { root } = result;
    const [fooRule] = root.nodes;
    expect((fooRule as any).selector).to.equal('.box-foo');
    const [colorDecl] = (fooRule as any).nodes;
    expect(colorDecl.value).to.be.a('string');
    expect(colorDecl.value).to.be.equal('undefined');

    expect(result.warnings().length).to.be.equal(1);
    expect(result.warnings()[0].type).to.be.equal('warning');
    expect(result.warnings()[0].text).to.be.equal(
      'variable --missing is undefined and used without a fallback'
    );
  });
});
