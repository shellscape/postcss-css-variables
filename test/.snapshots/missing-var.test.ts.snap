// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`missing-var > use fallback value if provided with missing variables 1`] = `".box-foo{width:undefined}.box-bar{width:30px}"`;

exports[`missing-var > use fallback variable if provided with missing variables 1`] = `".box-foo{width:undefined}.box-bar{width:100px}"`;

exports[`missing-var > use fallback variable if provided with missing variables calc 1`] = `".box-bar{width:calc(150px + 100px)}"`;

exports[`missing-var > use fallback variable if provided with missing variables nested 1`] = `".box-bar{width:100px}"`;

exports[`missing-var > work with missing variables 1`] = `".box-foo{width:undefined}"`;
